storage/innobase/xtrabackup/src/xtrabackup.cc

```diff
@@ -2156,11 +2156,94 @@ static bool innodb_init_param(void) {
   return (TRUE);
 }
 
+/** Scan the sdi id from mysql.tables
+@param[in]	name 		Name of table database/name
+@param[out]     sdi_id		id of table
+@param[in]	thd 		THD */
+void get_id_from_mysql_table_scan(std::string name, uint64 *id, THD *thd) {
+  std::string db_name = name.substr(0, name.find(OS_PATH_SEPARATOR));
+
+  std::string table_name = name.substr(name.find(OS_PATH_SEPARATOR) + 1);
+  if (table_name.find("#P#") != std::string::npos ||
+      table_name.find("#p#") != std::string::npos) {
+    table_name.resize(table_name.find('#'));
+  }
+  uint64 schema_id = 0;
+
+  ut_ad(db_name.compare("mysql") != 0);
+
+  dict_table_t *sys_tables = nullptr;
+  btr_pcur_t pcur;
+  const rec_t *rec = nullptr;
+  mtr_t mtr;
+  MDL_ticket *mdl = nullptr;
+  mem_heap_t *heap = mem_heap_create(1000);
+
+  mutex_enter(&(dict_sys->mutex));
+  mtr_start(&mtr);
+
+
+  rec = dd_startscan_system(thd, &mdl, &pcur, &mtr, "mysql/schemata",
+                            &sys_tables);
+  while (rec) {
+    MDL_ticket *mdl_on_tab = nullptr;
+    uint64 rec_schema_id;
+    std::string rec_name;
+
+    dd_process_schema_rec(heap, rec, sys_tables, &mdl_on_tab, &mtr, &rec_name,
+                          &rec_schema_id);
+    mem_heap_empty(heap);
+
+    if (rec_name.compare(db_name) == 0) {
+      schema_id = rec_schema_id;
+      break;
+    }
+    rec = (rec_t *)dd_getnext_system_rec(&pcur, &mtr);
+  }
```

Comment:
after the first scan, we should create the in-memory map and avoid future scans.

Reply:
I had this in mind. but given upstream has already fixed duplicate SDI and there won't be lot of duplicate just leftover from upgrade. If you still think it can be useful. I will create the map instead of scanning again and again.

Reply:
atleast for schemas, which are less in number compared to tables, we can do this.


For tables, am worried about memory usage. There can be like 1million entries. So not recommending at this point

Reply:
fixed